cannon: allocator preheap instrumentation [rebased from ec2#2]

salindne opened the pull request:
This PR revives [ec2#2](https://github.com/ec2/optimism/pull/2), originally opened by Eric from ChainSafe, for upstream consideration. It includes instrumentation for runtime allocator preheap behavior useful in constrained deployments of the OP Stack.

Changes were migrated to [ChainSafe/optimism#ec2/alloc-preheap](https://github.com/ChainSafe/optimism/tree/ec2/alloc-preheap) after Eric left our organization.

cc @clabby for context — reviving Eric's previous PR (https://github.com/ethereum-optimism/optimism/pull/2 on ec2/optimism) now ported to ChainSafe's fork here. Will try to resolve conflicts

ec2 committed:
Abstract the Merkle representation

ec2 committed:
Implement asterisc's MPT

ec2 committed:
migrate tests for mpt

ec2 committed:
copied benchmarks from asterisc

ec2 committed:
migrate tests for mpt

ec2 committed:
fix failed merge

ec2 committed:
Merge branch 'develop' into ec2/mem-abstraction

ec2 committed:
Merge branch 'develop' into ec2/mem-abstraction

ec2 committed:
Avoid pagelookup twice during setword invalidation

ec2 committed:
use uints

ec2 committed:
hashpool

ec2 committed:
alloc pages upfront

ec2 committed:
remove clutter

ec2 committed:
fix

ec2 committed:
gomallocregion

ec2 committed:
more readable mapped regions

ec2 committed:
Merge branch 'develop' into ec2/mem-abstraction

ec2 committed:
fix state json codec test

ec2 committed:
fix for singlethread too

ec2 committed:
fix op-challenger test

ec2 committed:
Merge branch 'develop' into ec2/mem-abstraction

ec2 committed:
Remove MPT implementation

ec2 committed:
address comments

ec2 committed:
fix benchmark

ec2 committed:
Use uint64 and also reuse hasher and buffers

ec2 committed:
Merge branch 'ec2/bitlist' into ec2/alloc-preheap

ec2 committed:
clean

ec2 committed:
clean up some

ec2 committed:
fix range

salindne commented:
Hi @clabby — reviving Eric's previous PR (#2 on `ec2/optimism`) now ported to ChainSafe's fork here. Will try to resolve conflicts

Inphi commented on cannon/mipsevm/arch/arch32.go:
```diff
@@ -0,0 +1,100 @@
+//go:build !cannon64
```
Merge conflict from the ec2 branch. MIPS32 is no longer supported, so we can delete this file. Actually this file doesn't exist in the develop branch. Please rebase.

Inphi commented on cannon/mipsevm/singlethreaded/state.go:
```diff
@@ -0,0 +1,329 @@
+package singlethreaded
```
This file was removed as the singlethreaded VM is no longer supported. Could you please rebase this PR on the latest develop?

Inphi reviewed (commented):
PR needs a rebase please.

ec2 committed:
Merge branch 'develop' into ec2/alloc-preheap

BlocksOnAChain commented:
@salindne can you address the latest comments we had, after our first round of reviews for this PR? 🙏

ec2 committed:
Merge branch 'develop' into ec2/alloc-preheap

ec2 committed:
Merge branch 'develop' into ec2/alloc-preheap

salindne commented:
We were able to rebase. From what I understand, the EFI/Kona suggestions by @clabby  were not implemented. Are you folks okay with merging  this work and then starting a new PR for those changes. @Inphi  @BlocksOnAChain

Inphi commented:
There's no need to make changes for Kona. PR scope looks good. I'll review this PR more carefully soon.

Inphi commented on cannon/mipsevm/memory/memory.go:
```diff
@@ -27,6 +27,27 @@ const (
+	return pageIndex >= m.start_addr>>PageAddrSize && pageIndex < m.end_addr>>PageAddrSize
+}
+
+func (m *MappedMemoryRegion) AccessWordBytes(addr Word) ([]byte, bool) {
```
this function is unused.

salindne replied:
Thanks! removed it!

Inphi commented on cannon/mipsevm/memory/memory.go:
```diff
@@ -165,7 +234,21 @@ func (m *Memory) GetWord(addr Word) Word {
+			indexAdjusted := pageIndex - region.start_addr>>PageAddrSize
+			if indexAdjusted*PageSize >= Word(currLen) {
+				region.Data = region.Data[:(pageIndex+1)*PageSize]
+			}
```
This if statement seems like a no-op to me. At least given how the region maps are constructed in `NewBinaryTreeMemory`, if we're here, then the region map must contain the `pageIndex` and so it fully covers the entire page.

The if statement can only be true if the region map was constructed incorrectly, with `end_addr-start_addr > len(Data)`. But even in this case, it's not clear to me why the `region.Data` slice is being truncated.

salindne replied:
Thanks for flagging, removed the no-op

salindne committed:
Merge branch 'ec2/alloc-preheap' of https://github.com/ChainSafe/optimism into ec2/alloc-preheap

Inphi commented on cannon/mipsevm/memory/binary_tree.go:
```diff
@@ -15,10 +17,23 @@ type BinaryTreeIndex struct {
+		Data:       make([]byte, 1<<31),
+	}
+	indexedRegions[1] = MappedMemoryRegion{
+		start_addr: arch.ProgramHeapStart,
```
nit: If `ProgramHeapStart > 1<31` thus altering the end address of the first indexed region (L24), then it wouldn't be appropriate to start at `archi.ProgramHeapStart`.
Given that the 32-bit VM is no longer used, I suggest simplifying this by hardcoding the exact start and end addresses of both regions. Rather than doing the `min` operation so it's quite obvious the bounds of the regions.

Inphi commented on cannon/mipsevm/memory/binary_tree.go:
```diff
@@ -15,10 +17,23 @@ type BinaryTreeIndex struct {
+	}
+	indexedRegions[1] = MappedMemoryRegion{
+		start_addr: arch.ProgramHeapStart,
+		end_addr:   min(arch.HeapStart, arch.ProgramHeapStart+(1<<31)),
```
nit: Similar to my comment on L28, simply hardcode the `end_addr` value.

Inphi commented on cannon/mipsevm/multithreaded/mips.go:
```diff
@@ -311,8 +311,8 @@ func (m *InstrumentedState) doMipsStep() error {
 
-	//instruction fetch
-	insn, opcode, fun := exec.GetInstructionDetails(m.state.GetPC(), m.state.Memory)
+	insn_detail := m.cached_decode[m.state.GetPC()/4]
```
The cache fits 512M instructions, which should always suffice. But in case it doesn't, let's fallback to reading and decoding the instruction from memory if the PC is larger than the cache.

Inphi commented on cannon/mipsevm/memory/memory.go:
```diff
@@ -27,6 +27,20 @@ const (
 type Word = arch.Word
 
+type MappedMemoryRegion struct {
+	start_addr Word
+	end_addr   Word
```
nit: camelCase for Go identifiers is the convention.
```suggestion
	startAddr Word
	endAddr   Word
```

Inphi reviewed (commented):
Changes look good. Let's add a fallback for cache misses and should be good to merge.

pauldowman commented:
@salindne this is so close, thank you! There are a couple of tiny bits of feedback, would you be able to address those and then we can merge? 🙏

salindne committed:
Address reviewer feedback: clarify region bounds and add instruction cache fallback

salindne commented:
I accidentally pushed a bunch of extra changes I didn't mean to, I need to revert and fix give me another day.

Sorry for the review tags

salindne committed:
Revert "Address reviewer feedback: clarify region bounds and add instruction cache fallback"

This reverts commit 2a19b224db0fec673117d2a52e75324fdf422db2.

salindne committed:
Clarify region bounds and add fallback to memory decode if PC exceeds instruction cache

salindne commented:
Reverted and then got the fixes in, thank you all

Inphi commented on cannon/mipsevm/multithreaded/mips.go:
```diff
@@ -311,8 +311,18 @@ func (m *InstrumentedState) doMipsStep() error {
+		rawInsn := m.state.Memory.GetWord(pc)
+		insn = uint32(rawInsn)
+		opcode = (insn >> 26) & 0x3F
+		fun = insn & 0x3F
```
```suggestion
	    insn, opcode, fun = exec.GetInstructionDetails(pc, m.state.Memory)
```
the decode function handles address masking. Let's use that for consistency.

Inphi reviewed (approved):
Excellent! Thanks!

salindne committed:
Update cannon/mipsevm/multithreaded/mips.go

Co-authored-by: Inphi <mlaw2501@gmail.com>

Inphi commented:
/ci authorize 2965ca1abaff87a669d5e35edf5b7be628713bb9

Inphi commented on cannon/mipsevm/memory/binary_tree.go:
```diff
@@ -15,10 +17,23 @@ type BinaryTreeIndex struct {
+	indexedRegions[0] = MappedMemoryRegion{
+		startAddr: 0,
+		endAddr:   arch.ProgramHeapStart,
+		Data:      make([]byte, arch.ProgramHeapStart),
```
The `arch.ProgramHeapStart` is too large to allocate. Shouldn't this be `1<<31`?

Inphi commented on cannon/mipsevm/memory/binary_tree.go:
```diff
@@ -15,10 +17,23 @@ type BinaryTreeIndex struct {
+	indexedRegions[1] = MappedMemoryRegion{
+		startAddr: arch.ProgramHeapStart,
+		endAddr:   arch.HeapStart,
+		Data:      make([]byte, arch.HeapStart-arch.ProgramHeapStart),
```
Same as my above comment.

Inphi reviewed (changes requested):
The changes are failing in [CI](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/95658/workflows/22bc3421-3f07-4734-987d-b911d3abc31a/jobs/3696743). I think fixing size of the preheap cache should fix. Worth running `go test ./cannon/...` to make sure nothing else is broken by this change.

salindne committed:
fixed size of preheap cache for binaryTreeMemory

salindne commented:
Sorry about that, was overthinking. I ran test and on cannon/mipsevm/memory and it passed after fix to 1<<31.

Inphi commented:
/ci authorize 05b870f0d589f9ad1e5ecf7c31796f15a5c19f8f

salindne commented:
Didnt realize go test ./cannon/... runs all tests, I see others have failed aswell, Ill try to get through them now

salindne committed:
slice out of bounds fixed in GetWord and AllocPage, pagetest syntax fix

salindne commented:
Thank you for your help inphi, I was having trouble getting testdata working, but locally I think I fixed the slice out of bounds problem and also the issue that caused go-lint to fail. These were the issues i saw affecting all of the required tests that failed.

Inphi commented:
/ci authorize d311e21f92125988a02ca6220d9ac917dedca423

Inphi commented on cannon/mipsevm/memory/memory.go:
```diff
@@ -212,6 +212,9 @@ func (m *Memory) GetWord(addr Word) Word {
 		if ok := region.AddrInRegion(addr); ok {
 			offset := addr - region.startAddr
+			if offset+arch.WordSizeBytes > Word(len(region.Data)) {
+				return 0 // out of bounds access, return zero
```
I don't think this is appropriate. If the address isn't in the preheap then we should fallback to reading from regular memory. There isn't any guarantee that the value at that address is empty.

Inphi replied:
Also, given the `region.AddrInRegion` check on L213, I wouldn't expect an out of bounds access to be possible. If you're seeing out of bounds accesses in the test then this implies the preheap isn't constructed correctly. Possibly something broke in the recent changes?

salindne replied:
Okay Ill remove these additions, the test failure I was trying to resolve is

```
--- FAIL: TestLoadSubWord_32bits (0.00s)
    --- FAIL: TestLoadSubWord_32bits/32-bit (0.00s)
panic: runtime error: slice bounds out of range [::4278194176] with capacity 2147483648 [recovered]
        panic: runtime error: slice bounds out of range [::4278194176] with capacity 2147483648

goroutine 8 [running]:
testing.tRunner.func1.2({0x1006447e0, 0x140000180f0})
        /opt/homebrew/Cellar/go/1.24.3/libexec/src/testing/testing.go:1734 +0x1ac
testing.tRunner.func1()
        /opt/homebrew/Cellar/go/1.24.3/libexec/src/testing/testing.go:1737 +0x334
panic({0x1006447e0?, 0x140000180f0?})
        /opt/homebrew/Cellar/go/1.24.3/libexec/src/runtime/panic.go:792 +0x124
github.com/ethereum-optimism/optimism/cannon/mipsevm/memory.(*Memory).AllocPage(0x14000148050, 0xff000)
        /Users/s3b/Dev/zkVerify/optimism/cannon/mipsevm/memory/memory.go:234 +0x1d8
github.com/ethereum-optimism/optimism/cannon/mipsevm/memory.(*Memory).SetWord(0x14000148050?, 0xff000000, 0x1234567800000000)
        /Users/s3b/Dev/zkVerify/optimism/cannon/mipsevm/memory/memory.go:194 +0x110
github.com/ethereum-optimism/optimism/cannon/mipsevm/exec.TestLoadSubWord_32bits.func1(0x140001588c0)
        /Users/s3b/Dev/zkVerify/optimism/cannon/mipsevm/exec/mips_instructions32_test.go:52 +0x25c
testing.tRunner(0x140001588c0, 0x140001116e0)
        /opt/homebrew/Cellar/go/1.24.3/libexec/src/testing/testing.go:1792 +0xe4
created by testing.(*T).Run in goroutine 7
        /opt/homebrew/Cellar/go/1.24.3/libexec/src/testing/testing.go:1851 +0x374
FAIL    github.com/ethereum-optimism/optimism/cannon/mipsevm/exec       0.196s
```
One thing I can think of right now is during preheap construction (binary_tree.go), in the mapped regions

```
func NewBinaryTreeMemory() *Memory {
	pages := make(map[Word]*CachedPage)
	index := NewBinaryTreeIndex(pages)

	indexedRegions := make([]MappedMemoryRegion, 2)
	indexedRegions[0] = MappedMemoryRegion{
		startAddr: 0,
		endAddr:   arch.ProgramHeapStart,
		Data:      make([]byte, 1<<31),
	}
	indexedRegions[1] = MappedMemoryRegion{
		startAddr: arch.ProgramHeapStart,
		endAddr:   arch.HeapStart,
		Data:      make([]byte, 1<<31),
	}
	return &Memory{
		merkleIndex:   index,
		pageTable:     pages,
		lastPageKeys:  [2]Word{^Word(0), ^Word(0)}, // default to invalid keys, to not match any pages
		MappedRegions: indexedRegions,
	}
}
```
The data is initialized to 2gib, but the regions are like 3000gib+, (from arch64.go)

```
const (
	AddressMask = 0xFFFFFFFFFFFFFFF8
	WordSize    = 64
	ExtMask     = 0x7

	// Ensure virtual address is limited to 48-bits as many user programs assume such to implement packed pointers
	Limit            = 0x00_00_FF_FF_FF_FF_FF_FF
	ProgramHeapStart = 0x00_00_00_c0_00_00_00_00
	HeapStart        = 0x00_00_10_00_00_00_00_00
	HeapEnd          = 0x00_00_60_00_00_00_00_00
	ProgramBreak     = 0x00_00_40_00_00_00_00_00
	HighMemoryStart  = 0x00_00_7F_FF_FF_FF_F0_00
)
```


I had originally made the data the same size, and I agree thats not the right call. Do you think discrepancies here is the problem?

salindne replied:
Got the slciing fixed, was just a dumb mistake when removing the 32vs 64 bit comapitbility, the following works (and makes sense):

```
func NewBinaryTreeMemory() *Memory {
	pages := make(map[Word]*CachedPage)
	index := NewBinaryTreeIndex(pages)

	indexedRegions := make([]MappedMemoryRegion, 2)
	indexedRegions[0] = MappedMemoryRegion{
		startAddr: 0,
		endAddr:   1 << 31,
		Data:      make([]byte, 1<<31),
	}
	indexedRegions[1] = MappedMemoryRegion{
		startAddr: arch.ProgramHeapStart,
		endAddr:   arch.ProgramHeapStart + 1<<31,
		Data:      make([]byte, 1<<31),
	}
```

Now all thats left when I test is multithreaded fails, I get failure on the test after  last changes
```

Random int: 11733101251531920949
Random hex data: 37da195c4bf8d4f1a3c00955d6512a1e8ae51e44412acad37f
Random hex data: 0533e6e1f6
signal: killed
FAIL    github.com/ethereum-optimism/optimism/cannon/mipsevm/multithreaded      94.436s
```

Inphi commented on cannon/mipsevm/memory/memory.go:
```diff
@@ -230,8 +233,15 @@ func (m *Memory) AllocPage(pageIndex Word) *CachedPage {
+			if end <= Word(len(region.Data)) && start < end { //start >= 0 not needed as start is unint and always >= 0
+				p.Data = region.Data[start:end:end]
+			} else {
+				p.Data = make(Page, PageSize)
```
Similar to my above comment, we should not be expecting an out of bounds access due to the check on L235.

Inphi reviewed (commented):
I'm seeing a couple failed tests - https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/95708/workflows/761a3c0e-49ff-442e-b3af-ea498c5a899b/jobs/3698983/steps. There seems to be a regression introduced recently. I haven't look carefully into the issue, but I suspect it's from the changes to the preheap setup.

I can dig into this a bit more tomorrow.

salindne committed:
fix program heap initialization for binary tree

salindne commented:
Im getting os signal kill and time out issues for multithread tests on my machine, don't know if its just my machine or not. I reverted the fallback change to mips from last week and had the same result, so I have left mips for now. All other tests are passing for me locally.

Inphi commented:
/ci authorize e243074181cbbcc9e75371406553b2478ba77345

Inphi commented:
@salindne Tests are failing in CI for two reasons:
1. Flakes in an older mainline commit you have merged. Merging the latest develop should fix these.
2. The cannon tests, [see this](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/96056/workflows/d8e1255f-d0fa-4c3a-b8ea-f5f6eefaee4b/jobs/3707356), are failing because the instruction cache is empty when those tests run. The cannon testing framework involves manually constructing a VM state to set up the condition we want to test. As an [example](https://github.com/ethereum-optimism/optimism/blob/ddcaebae57ab1ef731f82f3a0cfcc82cf070ff50/cannon/mipsevm/tests/evm_multithreaded_test.go#L62), we manually inject a specific instruction being tested into the VM state memory. The test then executes a single step to observe the VM post-state. However, the VM memory is zeroed by the time the instruction cache is constructed as we don't typically "load" programs in single-step unit tests. Thus, the instruction cache remains empty and the VM reads an invalid zero value when it decodes an instruction. So you would have to ensure that all instructions stubbed onto memory are reflected in the instruction cache.

There is another problem that this PR should address, which is the test execution time. Populating the instruction cache takes several seconds on my machine. While this isn't an issue in a real deployment as it's a tiny fraction of overall execution time, this is too long a time for unit and fuzz testing which typically execute in milliseconds.
Since the bulk of the time construction the instruction cache is iterating and decoding instructions in low memory (`state.Memory.MappedRegions`), I suggest resizing the mapped region area for unit tests. A region with a 128 instruction capacity more than suffices. Note that there are also tests that execute entire guest programs. Those tests should continue to use the default 2 GiB memory region size.

salindne commented:
Thank you for your feedback, it is very helpful.

salindne committed:
Merge remote-tracking branch 'origin/develop' into ec2/alloc-preheap
Merge branch 'develop' into ec2/preheap

Inphi commented:
/ci authorize 0b843b5c4548da91ad9b84cffc6632c739808261

salindne committed:
moved cache decoded from instrumented state declaration to doMipsStep

salindne committed:
added default and set region map size to newMemory

salindne committed:
added 4096 byte memory region option to all testing cases using VMFactory

salindne commented:
For second comment, moved cache decode logic from instrument initialization to mips step which seemed to fix the problem.

For testing memory region, pretty carefully went through all tests in the tests folder, added an additional option for VMFactory to set region size leaving standard calls as 2gib default. 

I have a couple tests around elf files that always fail for me, even develop branch, and then i had 2 multithread64 tests trigger signal kill from my OS that I cant confirm, otherwise everything else seems to pass. 

Finally, I know 128 instructions (512 bytes) was suggested for the memory region, but there were a lot of tests failing at the size, I tried 1024, 2048, and found all tests pass at 4096 and set it to that.

salindne committed:
clean up

Inphi commented:
/ci authorize a0e4f05c6bd6448e3f3da8fa5219e6cbd311454b

Inphi commented:
> For second comment, moved cache decode logic from instrument initialization to mips step which seemed to fix the problem.

Isn't the `cache_decode` an ordered set of decoded instructions? Thus, I think moving the logic to mips step breaks that ordering and makes it impossible to locate a cached instruction without searching the entire set.

salindne committed:
cache_decode to be initialize on first mipstep only for single step test compatibility

salindne committed:
added heap region sizing for unit/fuzz tests

salindne commented:
> Isn't the `cache_decode` an ordered set of decoded instructions? Thus, I think moving the logic to mips step breaks that ordering and makes it impossible to locate a cached instruction without searching the entire set.

You are right. 

Trying to make precompute cache work with single step test structure where mipstep doesn't fallback to grabbing from memory every step. I changed it to only grab from memory and populate the cache on first step (to work with single step testing), and then grabs from cache every subsequent step at the cost of an if statement within the step logic. Maybe theres a better way.

I also added the ability to adjust heap region size and applied it to VMFactory calls. Seems to help with speed of tests, I see Fuzz tests timing out in the last ci run.

Please let me know if you if any suggestions for improving test suite timing. 

go test ./cannon/...  works for me like develop branch other than its slower and

`github.com/ethereum-optimism/optimism/cannon/mipsevm/tests`

times out at 10 minutes.

Inphi commented:
I think the previous behavior where it pre-allocates the `cached_decode` slice is better. With the configurable memory region size, we can still pre-allocate the preheap quickly for unit testing. And indexing to a slice is much faster than using a map.
I haven't run the benchmarks but I suspect using a map here would undo the performance gains of having a preheap. Since reading instructions from VM memory is also mostly cost-dominated by accessing the page table map.

So adding the change at https://github.com/ethereum-optimism/optimism/pull/16030/commits/7d86885022808701a074daca777bb4e99cadcc5f on top of a0e4f05c6bd6448e3f3da8fa5219e6cbd311454b should do.

salindne commented:
Okay I thank you, reverting back would mean the cache will be stale for the single step unit tests. 

I wasnt sure if it was accepteble to change the tests/VMFactory.

Do you want me change the test files with adding extra slice initialization somewhere in the VMFactory/testing code. 

Edit: I missed your edit/ last line.  What about your comment

> Isn't the `cache_decode` an ordered set of decoded instructions? Thus, I think moving the logic to mips step breaks that ordering and makes it impossible to locate a cached instruction without searching the entire set.

salindne committed:
Revert " cache_decode to be initialize on first mipstep only for single step test compatibility"

This reverts commit 47a716645290a48e6093c776d2ad7fede37ce2af.

Inphi commented:
> Okay I thank you, reverting back would mean the cache will be stale for the single step unit tests.

Do you mean that the preheap with 4K elements isn't enough to hit the cache for single step unit tests? I think what might be happening is some PC randomization used by the test framework - https://github.com/ethereum-optimism/optimism/blob/0ea9f6cd2dde475fa6ef15b6ed2ad131fc53bb36/cannon/mipsevm/multithreaded/testutil/thread.go#L13. 
Many thread-specific tests rely on that utility to create random threading states. Try clipping the maximum random PC value according to the size of the memory region used for testing.

Inphi commented:
> Edit: I missed your edit/ last line. What about your comment
> 
> > Isn't the `cache_decode` an ordered set of decoded instructions? Thus, I think moving the logic to mips step breaks that ordering and makes it impossible to locate a cached instruction without searching the entire set.

Sorry I meant to suggest that you move rebuild the https://github.com/ethereum-optimism/optimism/commit/7d86885022808701a074daca777bb4e99cadcc5f change on top of https://github.com/ethereum-optimism/optimism/pull/16030/commits/0b843b5c4548da91ad9b84cffc6632c739808261

salindne commented:
Back to this comment:


> 2\. The cannon tests, [see this](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/96056/workflows/d8e1255f-d0fa-4c3a-b8ea-f5f6eefaee4b/jobs/3707356), are failing because the instruction cache is empty when those tests run. The cannon testing framework involves manually constructing a VM state to set up the condition we want to test. As an [example](https://github.com/ethereum-optimism/optimism/blob/ddcaebae57ab1ef731f82f3a0cfcc82cf070ff50/cannon/mipsevm/tests/evm_multithreaded_test.go#L62), we manually inject a specific instruction being tested into the VM state memory. The test then executes a single step to observe the VM post-state. However, the VM memory is zeroed by the time the instruction cache is constructed as we don't typically "load" programs in single-step unit tests. Thus, the instruction cache remains empty and the VM reads an invalid zero value when it decodes an instruction. So you would have to ensure that all instructions stubbed onto memory are reflected in the instruction cache.


Pre allocating to a slice during initialization of the InstrumentedState, was causing stale cache issues with single step tests.

The attempt to move cache allocation to the mips step, was inefficient as you mentioned, then trying to use a map is also inefficient as you pointed out. 

Going back to pre allocation using a slice in the instrumentedState initialization then reintroduces the original stale cache issue with the tests where instructions and registers were 0.

My question is do you think its appropriate to add a "reallocateCache" function to the instrumented state class and call that after memory injection, before step, in all of the tests? 

I am trying to figure out a way checking if the cache is stale in the doMipsStep function and acting accordingly. Not sure if this is a preferred way to go about or to leave doMipsStep as it was and change the tests everywhere.

I do think your suggestion about randomization needs to be looked at as well, and ill do that.

salindne commented:
> Sorry I meant to suggest that you move rebuild the [7d86885](https://github.com/ethereum-optimism/optimism/commit/7d86885022808701a074daca777bb4e99cadcc5f) change on top of [0b843b5](https://github.com/ethereum-optimism/optimism/commit/0b843b5c4548da91ad9b84cffc6632c739808261)

Okay, here I only reduced the size of the instruction sets, later I added reducing the size of the heap as well, which seemed to improve test time in a lot of cases. I assume you want both left in.

salindne committed:
cache allocation in instrumentedState initialization, catch for stale cache in doMipsStep

salindne commented:
Thank you for your help @Inphi. I pushed what you suggested:

-with reduced region sizes for tests, 
-preallocation of cache in NewInstrumentedState initialization
-doMipsStep acts as before, but I added a check that then allocates cache when instructions are 0, or when pc larger than cache size catching large random pc values in the tests for now. 

From here, let me know if you want me to go through all of the tests and reallocate cache before invoking step, and clip the random PC values in the tests, in order to remove the logic from doMipsStep.

Inphi commented:
/ci authorize 43f5bf1c726c75d01651309aee3ccc5222300a70

Inphi commented on cannon/mipsevm/multithreaded/mips.go:
```diff
@@ -321,8 +321,23 @@ func (m *InstrumentedState) doMipsStep() error {
+		if decoded.insn != 0 {
+			insn, opcode, fun = decoded.insn, decoded.opcode, decoded.fun
+		} else {
+			// Fallback decode for uninitialized slot
+			insn, opcode, fun = exec.GetInstructionDetails(pc, m.state.Memory)
+			m.cached_decode[cacheIdx] = InstructionDetails{insn, opcode, fun}
```
When `decoded.insn == 0`, this may indeed be a valid instruction (for `sll`). And since the entire `m.cached_decode` buffer is allocated, it's not possible for there to be uninitialized values in there.
So falling back here is unnecessary. It's OK to use the zero valued instruction as-is.

salindne replied:
Okay, I understand, unfortunately this logic is what enables single step tests to work. Are you saying to remove it, and then the singlestep tests can fail within the ci tests. 

The only reason the logic for checking insn!=0 is there is becuase for the single step/fuzz tests, after memory injection , the cache is stale and needs to be allocated.

I understand that decoded.insn == 0 may represent a valid instruction, this is why I was asking earlier if you want me to remove this logic from the doMipsStep, at the cost of patching all of the single step/fuzz tests with reallocation logic.

Edit: Removing line 335, and everything seems to work the same. Hopefully this is what you meant, and if not please let me know about my above question

Inphi replied:
Yes, I think it should be removed. I'll take a closer look at why the tests are failing without it. While it makes sense to have that fallback to accomodate tests, it's awkward to do so. There should be a better way to fix tests rather than have that contrivance in the VM.

salindne replied:
I agree it is akward, the only way I can think of is to add something like 

```
func (m *InstrumentedState) RefreshInstructionCache(startPC, endPC Word) {
	for pc := startPC; pc < endPC; pc += 4 {
		idx := pc / 4
		if int(idx) < len(m.cached_decode) {
			insn, opcode, fun := exec.GetInstructionDetails(pc, m.state.Memory)
			m.cached_decode[idx] = InstructionDetails{insn, opcode, fun}
		}
	}
}
```

to instrument.go

and then call this after memory injections. For example  lines 50-53 in evm_common_test.go:

```
goVm := v.VMFactory(nil, os.Stdout, os.Stderr, testutil.CreateLogger(), mtutil.WithRandomization(int64(i)), mtutil.WithPC(tt.pc), mtutil.WithNextPC(tt.nextPC), mtutil.WithRegionSize(testCodeRegionSize, testHeapSize))
state := goVm.GetState()
testutil.StoreInstruction(state.GetMemory(), tt.pc, tt.insn)
step := state.GetStep()
```

replaced with something like

```
goVm := v.VMFactory(nil, os.Stdout, os.Stderr, testutil.CreateLogger(), mtutil.WithRandomization(int64(i)), mtutil.WithPC(tt.pc), mtutil.WithNextPC(tt.nextPC), mtutil.WithRegionSize(testCodeRegionSize, testHeapSize))
state := goVm.GetState()
testutil.StoreInstruction(state.GetMemory(), tt.pc, tt.insn)

if instVm, ok := goVm.(*multithreaded.InstrumentedState); ok {
	instVm.RefreshInstructionCache(tt.pc, tt.pc+4)
}
step := state.GetStep()
```

in all of the tests that use VMFactory. 

Then the logic is scoped only to tests and wont affect normal computation. Not sure if this is the best way.

Inphi replied:
I think the best way to do this would be to move calls to `testutil.StoreInstruction` to the VMFactory. So that when the initial VM state is constructed, memory is also preset with the instruction used for the unit test. But I don't recommend doing this in this PR as it would be a big refactor.

For now, I think what you've suggested is OK. To reduce boilerplate, I suggest tweaking `testutil.StoreInstruction` so it updates the cache of the specific instruction being stored. This can be done as a single item update, rather than a cache refresh, so don't accidentally slowdown tests again.

Inphi reviewed (commented):
The changes look good. I just have one request on handling "uninitialized" instructions.
Should be good to merge once there are no issues running the change on a real op-program in CI.

salindne committed:
remove unit test logic from domipstep and added helper for cache update in tests

salindne commented:
When trying to add the cache updating into storeinstruction, the instrumented state has to be passed, and I came across a dependency cycle. I made a new package "helpers" in testutil, to get around this, defining 

```
func StoreInstructionWithCacheUpdate(mem *memory.Memory, pc arch.Word, insn uint32, vm any) {
	testutil.StoreInstruction(mem, pc, insn)
	if instVm, ok := vm.(*multithreaded.InstrumentedState); ok {
		instVm.UpdateInstructionCache(pc)
	}
}
```

and then replacing testutil.StoreInstruction with helpers.StoreInstructionWithCacheUpdate everywhere necessary.

salindne commented:
Looks like the tests are actively being changed to use difftester instead of vmfactory this week

https://github.com/ethereum-optimism/optimism/pull/16799

I am going to have to rework what was done to align with difftester

Inphi commented:
> Looks like the tests are actively being changed to use difftester instead of vmfactory this week
> 
> #16799
> 
> I am going to have to rework what was done to align with difftester

Sorry about that. There are a few more difftester changes that are being merged that may cause further conflicts. @mbaxter could you look into merging those PRs when you can?

mbaxter commented:
> @mbaxter could you look into merging those PRs when you can?

Yes, I will try to get these merged.  I can also help resolve the merge conflicts here.

salindne commented:
@mbaxter 

Thank you! Ill look through the difftester stuff and can probably get started to see if I can get some of the tests already migrated working

salindne committed:
merge with current develop

salindne committed:
update multithreaded test

salindne commented:
I jumped in a bit. The stale cache issues seem to have gone with the changed testing, but will still probably want to add limiting the instruction set and heap size to the tests.

Ill wait for the rest to get merge, looks like fuzz tests are left, and can probably add the code/heap size VMfactory options in the difftester.go file after and remove the StoreInstructionWithCacheUpdate function I added earlier that helped with the stale cache issues.

Will wait for confirmation that test migration is done. 

Thank you!

mbaxter commented:
>  looks like fuzz tests are left ... Will wait for confirmation that test migration is done.

Yes - just need to get [this last PR](https://github.com/ethereum-optimism/optimism/pull/16851) merged and then the migration will be done and tests will be stable again 🙏

mbaxter commented:
> Yes - just need to get [this last PR](https://github.com/ethereum-optimism/optimism/pull/16851) merged and then the migration will be done and tests will be stable again 🙏

Update - this PR is merged now!

salindne committed:
merge fuzz test migrations

Inphi commented:
> > Yes - just need to get [this last PR](https://github.com/ethereum-optimism/optimism/pull/16851) merged and then the migration will be done and tests will be stable again 🙏
> 
> Update - this PR is merged now!

Thanks. Let's rebase the migration changes from develop.

salindne committed:
merge migration changes from develop

Inphi commented:
/ci authorize 98ae85bd994dc13974aaf1121da5200f11663f6b

salindne committed:
some test step boundaries in instrumented_test needed to be slightly incremented, added decoded cache updateing to difftester_test.go

salindne commented:
Missed integrating difftester_test.go with changes to cache and instr/heap size. Some tests that failed in instrumented_test.go I got working locally after slightly increasing the max step limit.

Inphi commented:
/ci authorize 25b4751c0da398b37b24b2ef35eb48fe13b6e9e6

salindne committed:
small changes to step count and mismatch error vs string fault

Inphi commented:
/ci authorize 11b2176e19b29f1005d60d0c9214d440dc7bed78

salindne commented:
Im not sure how to approach the multithreaded test, looks like there is a timeout in the ci tests, the tests seem to individually pass, other than one test causes an OS signal kill for me locally.

For the fuzz tests, they all pass on my local implementation, Im not sure how to approach the error.

Inphi commented:
> Im not sure how to approach the multithreaded test, looks like there is a timeout in the ci tests, the tests seem to individually pass, other than one test causes an OS signal kill for me locally.
> 
> For the fuzz tests, they all pass on my local implementation, Im not sure how to approach the error.

The tests are OOM'ing in CI as the test runner doesn't have much memory to work with. Lemme figure something out..

Inphi committed:
cannon: Use tiny i-cache by default

* uses a small i-cache by default to accomodate unit tests. This
  prevents tests from running out of memory in CI.
* fixes fuzz tests to preload the i-cache with the test instruction.

Inphi committed:
op-e2e: Limit max parallelism for fp tests

The fp tests now require alot of memory to run. Particularly when cannon
is used. So limit parallelism to avoid OOM.

Inphi committed:
increase parallelism

Inphi commented:
@salindne Check out this branch that addresses the CI failures - https://github.com/ethereum-optimism/optimism/compare/inphi/ec2/alloc-preheap
Here's the diff of the changes for your convenience - https://github.com/ethereum-optimism/optimism/compare/11b2176e19b29f1005d60d0c9214d440dc7bed78..7707d3293bddfa5fd6030c5e1f2f66f7b8e1ad89.

The following adjustments were made to ensure that cannon is usable in CI:
- the binary heap is instantiated with an 8K memory region by default. Only when Cannon is executed via the `cannon/cmd` package would the full 4 GiB preheap be used to cache instructions.
- the test executor used by the `op-e2e/faultproofs` package prohibits too many tests from running in parallel.

If this all looks good to you, I can merge those changes into this PR and we should be good to merge to develop. I tried creating a PR on the Chainsafe/optimism targeting your branch but was unable to do so.

salindne commented:
@Inphi Looked through and looks good to me. Please do merge.

Inphi commented:
@salindne I don't have permission to push to your branch. I'm not sure why this didn't work before, but I've created a PR with the changes - https://github.com/ChainSafe/optimism/pull/10. Could you please merge that in.

salindne committed:
Merge pull request #10 from ethereum-optimism/inphi/ec2/alloc-preheap

cannon: Ensure FP tests are runnable in CI

Inphi commented:
/ci authorize ce9dc7e94c9c80b71d85f84dff62b20686b8f9e1
